kubernetes/config/kube_config.c

```diff
@@ -0,0 +1,309 @@
+    if (apiKeys) {
+
+        /* under development for the token based authentication */
+
+        *pApiKeys = apiKeys;
+    } else {
+        rc = -1;
+    }
+    return rc;
+}
+
+static char *getWorkingConfigFile(const char *configFileNamePassedIn)
+{
+    char *configFileName = NULL;
+    const char *kubeconfig_env = NULL;
+    const char *homedir_env = NULL;
+
+    if (configFileNamePassedIn) {
+        configFileName = strdup(configFileNamePassedIn);
+    } else {
+        kubeconfig_env = getenv(ENV_KUBECONFIGE);
+        if (kubeconfig_env) {
+            configFileName = strdup(kubeconfig_env);
+        } else {
+            homedir_env = getenv(ENV_HOME);
+            if (homedir_env) {
+                configFileName = calloc(strlen(homedir_env) + strlen(KUBE_CONFIG_DEFAULT_LOCATION) + 1, sizeof(char));
+                if (configFileName) {
+                    sprintf(configFileName, KUBE_CONFIG_DEFAULT_LOCATION, homedir_env);
+                }
+            }
+        }
+    }
+
+    return configFileName;
+}
+
+static kubeconfig_context_t *kubeconfig_get_current_context(kubeconfig_context_t ** contexts, int contexts_count, const char *current_context_str)
+{
+    kubeconfig_context_t *current_context = NULL;
+
+    if (NULL == contexts || NULL == current_context_str) {
+        return NULL;
+    }
+
+    int i = 0;
+    for (i = 0; i < contexts_count; i++) {
+        if (0 == strcmp(contexts[i]->name, current_context_str)) {
+            current_context = contexts[i];
+            break;
+        }
+    }
+
+    return current_context;
+}
+
+static kubeconfig_cluster_t *kubeconfig_get_current_cluster(kubeconfig_cluster_t ** clusters, int clusters_count, const char *cluster_name)
+{
+    kubeconfig_cluster_t *current_cluster = NULL;
+
+    if (NULL == clusters || NULL == cluster_name) {
+        return NULL;
+    }
+
+    int i = 0;
+    for (i = 0; i < clusters_count; i++) {
+        if (0 == strcmp(clusters[i]->name, cluster_name)) {
+            current_cluster = clusters[i];
+            break;
+        }
+    }
+
+    return current_cluster;
+}
+
+static kubeconfig_user_t *kubeconfig_get_current_user(kubeconfig_user_t ** users, int users_count, const char *user_name)
+{
+    kubeconfig_user_t *current_user = NULL;
+
+    if (NULL == users || NULL == user_name) {
+        return NULL;
+    }
+
+    int i = 0;
+    for (i = 0; i < users_count; i++) {
+        if (0 == strcmp(users[i]->name, user_name)) {
+            current_user = users[i];
+            break;
+        }
+    }
+
+    return current_user;
+}
+
+int load_kube_config(char **pBasePath, sslConfig_t ** pSslConfig, list_t ** pApiKeys, const char *configFileName)
+{
+    static char fname[] = "load_kube_config()";
+
+    int rc = 0;
+
+    kubeconfig_context_t *current_context = NULL;
+    kubeconfig_cluster_t *current_cluster = NULL;
+    kubeconfig_user_t *current_user = NULL;
+
+    kubeconfig_t *kubeconfig = kubeconfig_create();
```

Comment:
Does this need to be heap allocated? If you stack allocated it, you could remove all of the `goto end` statements, and just `return`

Reply:
The structure ```kubeconfig_t``` (and its sub-fields such as kubeconfig_cluster_t, kubeconfig_context_t,etc...) incluces lots fields of  ```char *```, even though  kubeconfig_t (and its sub-fields such as kubeconfig_cluster_t, kubeconfig_context_t,etc...) is allocated in stack,  the ```char *``` fields are still allocated in heap. So we need call the function ```kubeconfig_free``` to free all memory before function returns.

Changing the  ```char *``` to ```char xxx[buffer_size]``` is also not a better solution.